compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -260,6 +261,11 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             _ => (false, None),
         };
 
+        let mut v = ParamFinder { params: vec![] };
+        // Get all type parameters from the predicate. If the predicate references a type parameter
+        // at all, then we can only suggestion a bound on an item that has access to that parameter.
+        v.visit_predicate(UpcastFrom::upcast_from(trait_pred, self.tcx));
```

Comment:
just use `.upcast()`.

Reply:
Also, what do you mean? We should never be suggesting a bound involving a parameter that we don't have "access to" --we shouldn't be checking predicates involving parameters that aren't in scope anyways?

Reply:
We currently [have a fallback](https://github.com/rust-lang/rust/blob/a31c5bfb6427afd162f554e48711829bf100f40f/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs#L456-L483) where regardless of what the predicate that failed, we end up suggesting a bound ([with weasel wording that there might be a better way](https://github.com/rust-lang/rust/blob/5b542866400ad4a294f468cfa7e059d95c27a079/tests/ui/associated-types/issue-38821.stderr#L14-L17)). This is in place because there are [some](https://github.com/rust-lang/rust/blob/41e8d152dc5abb5a706999ada8b059d3420af8f3/tests/ui/associated-types/issue-85103-layout-debug.stderr) [complex](https://github.com/rust-lang/rust/blob/41e8d152dc5abb5a706999ada8b059d3420af8f3/tests/ui/suggestions/restrict-type-not-param.stderr) ([if uncommon](https://github.com/rust-lang/rust/blob/41e8d152dc5abb5a706999ada8b059d3420af8f3/tests/ui/associated-types/issue-85103-layout-debug.stderr)) bounds that are still useful to have that were given only by that. But in #104089, we had a report of that fallback doing the wrong thing (obvious in retrospect). The predicate might be something like `Self: From<T>`, where the `T` is a param *only in the associated item* and not at the `trait`/`impl` level. That's what this additional check guards against: If any part of the predicate references one of the generics of that item, we don't skip suggesting the bound at that level because it would be invalid in the parent item.